sphinxcontrib/confluencebuilder/builder.py

```diff
@@ -323,6 +328,20 @@ def write_doc(self, docname, doctree):
                 navnode.bottom = True
                 doctree.append(navnode)
 
+        if self.add_pagesecnumbers or self.add_secnumbers:
+            # Add section numbers from toctree to builder
+            self.secnumbers = self.env.toc_secnumbers.get(docname, {})
+
+        # Add section number to page
+        if self.add_pagesecnumbers:
+            doctitle = self._parse_doctree_title(docname, doctree)
+            if self.secnumbers.get(''):
+                numbers = self.secnumbers['']
+                doctitle = '.'.join(map(str, numbers)) + self.secnumber_suffix + doctitle
+                doctitle = ConfluenceState.registerTitle(docname, doctitle,
+                    self.config.confluence_publish_prefix,
+                    self.config.confluence_publish_postfix)
```

Comment:
I am a bit concerned about this placement of this logic. From my (limited) understanding of the Sphinx engine, a builder's `write_doc` call can be invoked in parallel (especially since we explicitly set `allow_parallel` on the builder). With this logic, you are changing the title name for a `docname` which other documents may have been processed and using older references.

I wonder to address this is as simple as moving this logic into the `prepare_writing` stage; for example:
```
...

doctitle = self._parse_doctree_title(docname, doctree)
if not doctitle:
    continue

if self.add_pagesecnumbers or self.add_secnumbers:
    self.secnumbers = self.env.toc_secnumbers.get(docname, {})

if self.add_pagesecnumbers and self.secnumbers.get(''):
    nums = self.secnumbers['']
    suffix = self.secnumber_suffix
    doctitle = '.'.join(map(str, nums)) + suffix + doctitle

doctitle = ConfluenceState.registerTitle(docname, doctitle,
    self.config.confluence_publish_prefix,
    self.config.confluence_publish_postfix)
    
...
```
This assumes that `self.env.toc_secnumbers` is properly populated at this stage. This change also avoid unnecessarily registered the title twice.

Reply:
I solved it in a slightly different way. I moved the page numbering to `prepare_writing` as you suggested, but I kept the part that called `self.secnumbers = self.env.toc_secnumbers.get(docname, {})` in `write_doc`, as it needs to be done for every doc.
If I didn't, the section numbers available in the writing phase would always be the section numbers of the last processed document.